README.rst
Grid Layout widgets for trame
==================================

trame-grid-layout extend trame **widgets** with components that can be used to create some dynamic grid layout containers.
It leverage [vue-grid-layout](https://github.com/jbaysolutions/vue-grid-layout) which is a grid layout system, like 
[Gridster](http://dsmorse.github.io/gridster.js/), for Vue.js. Heavily inspired by [React-Grid-Layout](https://github.com/react-grid-layout/react-grid-layout).


Installing
-----------------------------------------------------------

trame-grid-layout can be installed with `pip <https://pypi.org/project/trame-grid-layout/>`_:

.. code-block:: bash

    pip install --upgrade trame-grid-layout


Usage
-----------------------------------------------------------

The `Trame Tutorial <https://kitware.github.io/trame/docs/tutorial.html>`_ is the place to go to learn how to use the library and start building your own application.

The `API Reference <https://trame.readthedocs.io/en/latest/index.html>`_ documentation provides API-level documentation.


License
-----------------------------------------------------------

trame-grid-layout is made available under the MIT License License. For more details, see `LICENSE <https://github.com/Kitware/trame-grid-layout/blob/master/LICENSE>`_
This license has been chosen to match the one use by `vue-grid-layout <https://github.com/react-grid-layout/react-grid-layout/blob/master/LICENSE>`_ which can be exposed via this library.


Community
-----------------------------------------------------------

`Trame <https://kitware.github.io/trame/>`_ | `Discussions <https://github.com/Kitware/trame/discussions>`_ | `Issues <https://github.com/Kitware/trame/issues>`_ | `RoadMap <https://github.com/Kitware/trame/projects/1>`_ | `Contact Us <https://www.kitware.com/contact-us/>`_

.. image:: https://zenodo.org/badge/410108340.svg
    :target: https://zenodo.org/badge/latestdoi/410108340


Enjoying trame?
-----------------------------------------------------------

Share your experience `with a testimonial <https://github.com/Kitware/trame/issues/18>`_ or `with a brand approval <https://github.com/Kitware/trame/issues/19>`_.


Development: Grabbing client before push to PyPI
-----------------------------------------------------------

To update the client code, run the following command line while updating the targeted version

.. code-block:: console

    mkdir -p ./trame_grid_layout/module/serve
    curl https://unpkg.com/vue-grid-layout@2.4.0 -Lo ./trame_grid_layout/module/serve/vue-grid-layout.js

Simple example
-----------------------------------------------------------

.. code-block:: python

    from trame.app import get_server
    from trame.ui.vuetify import SinglePageLayout
    from trame.widgets import vuetify, grid

    server = get_server()
    state = server.state

    LAYOUT = [
        {"x": 0, "y": 0, "w": 2, "h": 2, "i": "0"},
        {"x": 2, "y": 0, "w": 2, "h": 4, "i": "1"},
        {"x": 4, "y": 0, "w": 2, "h": 5, "i": "2"},
        {"x": 6, "y": 0, "w": 2, "h": 3, "i": "3"},
        {"x": 8, "y": 0, "w": 2, "h": 3, "i": "4"},
        {"x": 10, "y": 0, "w": 2, "h": 3, "i": "5"},
        {"x": 0, "y": 5, "w": 2, "h": 5, "i": "6"},
        {"x": 2, "y": 5, "w": 2, "h": 5, "i": "7"},
        {"x": 4, "y": 5, "w": 2, "h": 5, "i": "8"},
        {"x": 6, "y": 3, "w": 2, "h": 4, "i": "9"},
        {"x": 8, "y": 4, "w": 2, "h": 4, "i": "10"},
        {"x": 10, "y": 4, "w": 2, "h": 4, "i": "11"},
        {"x": 0, "y": 10, "w": 2, "h": 5, "i": "12"},
        {"x": 2, "y": 10, "w": 2, "h": 5, "i": "13"},
        {"x": 4, "y": 8, "w": 2, "h": 4, "i": "14"},
        {"x": 6, "y": 8, "w": 2, "h": 4, "i": "15"},
        {"x": 8, "y": 10, "w": 2, "h": 5, "i": "16"},
        {"x": 10, "y": 4, "w": 2, "h": 2, "i": "17"},
        {"x": 0, "y": 9, "w": 2, "h": 3, "i": "18"},
        {"x": 2, "y": 6, "w": 2, "h": 2, "i": "19"},
    ]

    with SinglePageLayout(server) as layout:
        layout.title.set_text("Grid layout")
        with layout.content:
            with grid.GridLayout(
                layout=("layout", LAYOUT),
                row_height=20,
            ):
                grid.GridItem(
                    "{{ item.i }}",
                    v_for="item in layout",
                    key="item.i",
                    v_bind="item",
                    classes="pa-4",
                    style="border: solid 1px #333; background: rgba(128, 128, 128, 0.5);",
                )

    if __name__ == "__main__":
        server.start()


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.3 (2023-04-25)
### Fix
* **api:** Expose only meaningful classes ([`0085faa`](https://github.com/Kitware/trame-grid-layout/commit/0085faae5e44a89216c73d902612fb7918a46064))

## v1.0.2 (2023-02-23)
### Fix
* **version:** Add __version__ ([`6a33f8a`](https://github.com/Kitware/trame-grid-layout/commit/6a33f8a3306697a5efa32aa68150c67ec2759602))

## v1.0.1 (2022-12-05)
### Fix
* **update:** Re-evaluate siblings on update ([`a79889c`](https://github.com/Kitware/trame-grid-layout/commit/a79889c6a971f30fbabbd9a074b9b8ed2c4d8cec))

## v1.0.0 (2022-09-13)
### Feature
* **layout:** Add update method on GridLayout ([`cda67a6`](https://github.com/Kitware/trame-grid-layout/commit/cda67a621e6ef9f37ce2c6e06a34594085772536))

### Breaking
* new API with implicit attribute handling  ([`cda67a6`](https://github.com/Kitware/trame-grid-layout/commit/cda67a621e6ef9f37ce2c6e06a34594085772536))

## v0.0.2 (2022-09-13)
### Fix
* **attributes:** Force some attributes to always evaluate ([`c19c266`](https://github.com/Kitware/trame-grid-layout/commit/c19c2666837b2977f9340d06789224d3abe83233))

### Documentation
* **changelog:** Fix spelling ([`d62cf54`](https://github.com/Kitware/trame-grid-layout/commit/d62cf543e0e24e04128fb67850a295e6fa06b312))

## v0.0.1 (2022-09-09)
### Fix
* **initial:** Initial version exposing vue-grid-layout ([`bf77a8e`](https://github.com/Kitware/trame-grid-layout/commit/bf77a8e56f6701978c66ceffd84a4c7ad2a56a0a))

### Documentation
* **README:** Update content ([`6297981`](https://github.com/Kitware/trame-grid-layout/commit/62979816e8c803aff3c0e3d292e4b8cb9cc269b2))
* **README:** Update content ([`10f0b22`](https://github.com/Kitware/trame-grid-layout/commit/10f0b22dfa1236e60e689bbb50dff014bf3966a0))


tests/test_import.py
def test_import():
    from trame_grid.widgets.grid import GridLayout  # noqa: F401
    from trame.widgets.grid import GridLayout  # noqa: F401,F811


trame_grid/__init__.py
from trame_client.utils.version import get_version

__version__ = get_version("trame-grid-layout")

ITEM_KEYS = [
    "x",
    "y",
    "w",
    "h",
    "min_w",
    "min_h",
    "max_w",
    "max_h",
    "is_draggable",
    "is_resizable",
    "is_bounded",
    "static",
    "drag_ignore_from",
    "drag_allow_from",
    "resize_ignore_from",
    "preserve_aspect_ratio",
    "drag_option",
    "resize_option",
]


def update_layout_item(dest, updates):
    for key in ITEM_KEYS:
        if key in updates and dest[key] != updates[key]:
            dest[key] = updates[key]


def update_layout(original, new_layout):
    item_map = {}
    id_to_delete = set()

    # index for fast lookup
    for item in original:
        n_id = item.get("i")
        item_map[n_id] = item
        id_to_delete.add(n_id)

    # Iterate new values
    for item in new_layout:
        n_id = item.get("i")
        id_to_delete.discard(n_id)
        if n_id in item_map:
            update_layout_item(item_map[n_id], item)
        else:
            original.append(item)

    # Remove any nodes in id_to_delete
    for n_id in id_to_delete:
        original.remove(item_map[n_id])


trame_grid/widgets/grid.py
"""Grid Widgets only support vue2.
"""
from trame_client.widgets.core import AbstractElement
from trame_client.widgets.html import Div
from .. import module, update_layout

__all__ = [
    "GridLayout",
    "GridItem",
]


class HtmlElement(AbstractElement):
    def __init__(self, _elem_name, children=None, **kwargs):
        super().__init__(_elem_name, children, **kwargs)
        if self.server:
            self.server.enable_module(module)


class GridLayout(HtmlElement):
    def __init__(self, children=None, layout_updated=None, **kwargs):
        """
        `GridLayout <https://jbaysolutions.github.io/vue-grid-layout/guide/properties.html#gridlayout>`_

        :param layout: The value must be an Array of Object items. Each item must have i, x, y, w and h properties. Please refer to the documentation for GridItem below for more information.
        :param responsive_layouts: This is the initial layouts of the grid per breakpoint if responsive is set to true. The keys of the Object are breakpoint names and each value is an Array of Object items as defined by layout prop. eg:{ lg:[layout items], md:[layout items] }. Setting the prop after the creation of the GridLayout has no effect. See also responsive, breakpoints and cols.
        :param col_num: (default: 12) Says how many columns the grid has. The value should be a natural number.
        :param row_height: (default: 150) Says what is a height of a single row in pixels.
        :param max_rows: (default: Infinity) Says what is a maximal number of rows in the grid.
        :param margin: (default: [10, 10]) Says what are the margins of elements inside the grid. The value must be a two-element Array of Number. Each value is expressed in pixels. The first element is a margin horizontally, the second element is a vertical margin.
        :param is_draggable: (default: True) Says if the grids items are draggable.
        :param is_resizable: (default: True) Says if the grids items are resizable.
        :param is_mirrored: (default: False) Says if the RTL/LTR should be reversed.
        :param is_bounded: (default: False) Says if the grid items are bounded to the container when dragging
        :param auto_size: (default: True) Says if the container height should swells and contracts to fit contents.
        :param vertical_compact: (default: True) Says if the layout should be compact vertically.
        :param restore_on_drag: (default: False) Says if the moved grid items should be restored after an item has been dragged over.
        :param prevent_collision: (default: False) Says if grid items will move when being dragged over.
        :param use_css_transforms: (default: True) Says if the CSS transition-property: transform; should be used.
        :param responsive: (default: False) Says if the layout should be responsive to window width. See also responsiveLayouts, breakpoints and cols.
        :param breakpoints: (default: "{ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }")  Breakpoints defined for responsive layout, the parameter represents the width of different devices:lg(large), md(medium), sm(small), xs(extra small). Sets widths on which column number changes. See also responsiveLayouts and cols.
        :param cols: (default: "{ lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 }") Defines number of columns for each breakpoint
        :param use_style_cursor: (default: True) Says if set the styleCursor option to true. When dragging freezes, setting this value to false may alleviate problems. This property is not reactive.
        :param transform_scale: (default: 1) Sets a scaling factor to the size of the grid items, 1 is 100%

        Events

        :param layout_created: Emitted on the component created lifecycle hook
        :param layout_before_mount: Emitted on the component beforeMount lifecycle hook
        :param layout_mounted: Emitted on the component mounted lifecycle hook
        :param layout_ready: Emitted when all the operations on the mount hook finish
        :param layout_updated: Every time the layout has finished updating and positions of all grid-items are recalculated
        :param breakpoint_changed: Every time the breakpoint value changes due to window resize

        """
        if layout_updated is None:
            layout_updated = (self.update, "[$event]")

        self._layout_name = kwargs.get("layout", kwargs.get("layout_sync"))
        if isinstance(self._layout_name, (tuple, list)):
            self._layout_name = self._layout_name[0]

        # Need to re-evaluate the siblings
        self._dirty_layout_key = f"{self._layout_name}_dirty_ts"
        Div(f"{{{{{ self._dirty_layout_key }}}}}", style="display: none;")

        super().__init__(
            "grid-layout",
            children,
            layout_updated=layout_updated,
            **kwargs,
        )
        self._attr_names += [
            ("layout", ":layout"),
            ("layout_sync", ":layout.sync"),
            ("responsive_layouts", "responsiveLayouts"),
            ("col_num", ":colNum"),
            ("row_height", ":rowHeight"),
            ("max_rows", ":maxRows"),
            ("margin", ":margin"),
            ("is_draggable", "isDraggable"),
            ("is_resizable", "isResizable"),
            ("is_mirrored", "isMirrored"),
            ("is_bounded", "isBounded"),
            ("auto_size", "autoSize"),
            ("vertical_compact", "verticalCompact"),
            ("restore_on_drag", "restoreOnDrag"),
            ("prevent_collision", "preventCollision"),
            ("use_css_transforms", "useCssTransforms"),
            ("responsive"),
            ("breakpoints"),
            ("cols"),
            ("use_style_cursor", "useStyleCursor"),
            ("transform_scale", "transformScale"),
        ]
        self._event_names += [
            ("layout_created", "layout-created"),
            ("layout_before_mount", "layout-before-mount"),
            ("layout_mounted", "layout-mounted"),
            ("layout_ready", "layout-ready"),
            ("layout_updated", "layout-updated"),
            ("breakpoint_changed", "breakpoint-changed"),
        ]

        self._server.state[self._dirty_layout_key] = 0

    def update(self, new_layout):
        if self._layout_name:
            # Do not mark self._layout_name dirty here or it will break the client
            update_layout(self._server.state[self._layout_name], new_layout)
            # Trigger an update for the div
            self._server.state[self._dirty_layout_key] += 1


class GridItem(HtmlElement):
    def __init__(self, children=None, **kwargs):
        """
        `GridItem <https://jbaysolutions.github.io/vue-grid-layout/guide/properties.html#griditem>`_


        :param i:
        :param x:
        :param y:
        :param w:
        :param h:
        :param min_w:
        :param min_h:
        :param max_w:
        :param max_h:
        :param is_draggable:
        :param is_resizable:
        :param is_bounded:
        :param static:
        :param drag_ignore_from:
        :param drag_allow_from:
        :param resize_ignore_from:
        :param preserve_aspect_ratio:
        :param drag_option:
        :param resize_option:

        Events

        :param move: Every time an item is being moved and changes position.

            .. code-block:: javascript

                function moveEvent(i, newX, newY){
                    console.log("MOVE i=" + i + ", X=" + newX + ", Y=" + newY);
                }

        :param resize: Every time an item is being resized and changes size

            .. code-block:: javascript

                resizeEvent: function(i, newH, newW, newHPx, newWPx){
                    console.log("RESIZE i=" + i + ", H=" + newH + ", W=" + newW + ", H(px)=" + newHPx + ", W(px)=" + newWPx);
                }

        :param resized: Every time an item is finished being resized and changes size

            .. code-block:: javascript

                /**
                 *
                 * @param i the item id/index
                 * @param newH new height in grid rows
                 * @param newW new width in grid columns
                 * @param newHPx new height in pixels
                 * @param newWPx new width in pixels
                 *
                 */
                resizedEvent: function(i, newH, newW, newHPx, newWPx){
                    console.log("RESIZED i=" + i + ", H=" + newH + ", W=" + newW + ", H(px)=" + newHPx + ", W(px)=" + newWPx);
                }

        :param container_resized: Every time the grid item/layout container changes size (browser window or other)

            .. code-block:: javascript

                /**
                 *
                 * @param i the item id/index
                 * @param newH new height in grid rows
                 * @param newW new width in grid columns
                 * @param newHPx new height in pixels
                 * @param newWPx new width in pixels
                 *
                 */
                containerResizedEvent: function(i, newH, newW, newHPx, newWPx){
                    console.log("CONTAINER RESIZED i=" + i + ", H=" + newH + ", W=" + newW + ", H(px)=" + newHPx + ", W(px)=" + newWPx);
                },

        :param moved: Every time an item is finished being moved and changes position

            .. code-block:: javascript

                movedEvent: function(i, newX, newY){
                    console.log("MOVED i=" + i + ", X=" + newX + ", Y=" + newY);
                }
        """
        super().__init__(
            "grid-item",
            children,
            **kwargs,
        )
        self._attr_names += [
            "i",
            "x",
            "y",
            "w",
            "h",
            ("min_w", "minW"),
            ("min_h", "minH"),
            ("max_w", "maxW"),
            ("max_h", "maxH"),
            ("is_draggable", "isDraggable"),
            ("is_resizable", "isResizable"),
            ("is_bounded", "isBounded"),
            "static",
            ("drag_ignore_from", "dragIgnoreFrom"),
            ("drag_allow_from", "dragAllowFrom"),
            ("resize_ignore_from", "resizeIgnoreFrom"),
            ("preserve_aspect_ratio", "preserveAspectRatio"),
            ("drag_option", "dragOption"),
            ("resize_option", "resizeOption"),
        ]
        self._event_names += [
            ("resize", "resize"),
            ("move", "move"),
            ("resized", "resized"),
            ("container_resized", "container-resized"),
            ("moved", "moved"),
        ]


examples/simple_grid.py
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import grid

server = get_server()
state = server.state

LAYOUT = [
    {"x": 0, "y": 0, "w": 2, "h": 2, "i": "0"},
    {"x": 2, "y": 0, "w": 2, "h": 4, "i": "1"},
    {"x": 4, "y": 0, "w": 2, "h": 5, "i": "2"},
    {"x": 6, "y": 0, "w": 2, "h": 3, "i": "3"},
    {"x": 8, "y": 0, "w": 2, "h": 3, "i": "4"},
    {"x": 10, "y": 0, "w": 2, "h": 3, "i": "5"},
    {"x": 0, "y": 5, "w": 2, "h": 5, "i": "6"},
    {"x": 2, "y": 5, "w": 2, "h": 5, "i": "7"},
    {"x": 4, "y": 5, "w": 2, "h": 5, "i": "8"},
    {"x": 6, "y": 3, "w": 2, "h": 4, "i": "9"},
    {"x": 8, "y": 4, "w": 2, "h": 4, "i": "10"},
    {"x": 10, "y": 4, "w": 2, "h": 4, "i": "11"},
    {"x": 0, "y": 10, "w": 2, "h": 5, "i": "12"},
    {"x": 2, "y": 10, "w": 2, "h": 5, "i": "13"},
    {"x": 4, "y": 8, "w": 2, "h": 4, "i": "14"},
    {"x": 6, "y": 8, "w": 2, "h": 4, "i": "15"},
    {"x": 8, "y": 10, "w": 2, "h": 5, "i": "16"},
    {"x": 10, "y": 4, "w": 2, "h": 2, "i": "17"},
    {"x": 0, "y": 9, "w": 2, "h": 3, "i": "18"},
    {"x": 2, "y": 6, "w": 2, "h": 2, "i": "19"},
]

with SinglePageLayout(server) as layout:
    layout.title.set_text("Grid layout")
    layout.toolbar.height = 32
    with layout.content:
        with grid.GridLayout(
            layout=("layout", LAYOUT),
            row_height=20,
        ):
            grid.GridItem(
                "{{ item.i }}",
                v_for="item in layout",
                key="item.i",
                v_bind="item",
                classes="pa-4",
                style="border: solid 1px #333; background: rgba(128, 128, 128, 0.5); touch-action: none;",
            )


if __name__ == "__main__":
    server.start()
